azurelinuxagent/ga/policy/policy_engine.py

```diff
@@ -41,15 +73,176 @@ def _log_policy(cls, msg, is_success=True, op=WALAEventOperation.Policy, send_ev
+                template["extensionPolicies"]["allowListedExtensionsOnly"] = allowlist_policy
+
+            # Validate and update template with global signature policy
+            signature_policy = extension_policies.get("signatureRequired")
+            if signature_policy is not None:
+                __validate_attribute_type("signatureRequired", signature_policy, bool, "extensionPolicies")
+                template["extensionPolicies"]["signatureRequired"] = signature_policy
+
+            # Parse individual extension policies
+            extensions = extension_policies.get("extensions")
+            if extensions is not None:
+                __validate_attribute_type("extensions", extensions, dict)
+                for ext_name, ext_policy in extensions.items():
+                    __validate_attribute_type(ext_name, ext_policy, dict, "extensions")
+                    individual_signing_policy = ext_policy.get("signatureRequired")
+                    if individual_signing_policy is not None:
+                        __validate_attribute_type("signatureRequired", individual_signing_policy, bool, ext_name)
+                    else:
+                        # If individual extension is present but signature policy not specified, use global policy.
+                        individual_signing_policy = template.get("extensionPolicies").get("signatureRequired")
+
+                    # Build individual extension policy and add to template.
+                    policy_to_add = {
+                        "signatureRequired": individual_signing_policy
+                    }
+                    template["extensionPolicies"]["extensions"][ext_name] = policy_to_add
+
+                # Convert "extensions" to a case-folded dict for case-insensitive lookup
+                case_folded_extension_dict = _CaseFoldedDict.from_dict(template["extensionPolicies"]["extensions"])
+                template["extensionPolicies"]["extensions"] = case_folded_extension_dict
+
+        return template
+
+
+class ExtensionPolicyEngine(_PolicyEngine):
+
+    def should_allow_extension(self, extension_to_check):
+        """
+        Return whether we should allow extension download based on policy.
+        extension_to_check is expected to be an Extension object.
+
+        If policy feature not enabled, return True.
+        If allowListedExtensionsOnly=true, return true only if extension present in "extensions" allowlist.
+        If allowListedExtensions=false, return true always.
+        """
+        if not self.policy_enforcement_enabled:
+            return True
+
+        allow_listed_extension_only = self._policy.get("extensionPolicies").get("allowListedExtensionsOnly")
+        extension_allowlist = self._policy.get("extensionPolicies").get("extensions")
+
+        should_allow = not allow_listed_extension_only or extension_allowlist.get(extension_to_check.name) is not None
```

Comment:
is extension_to_check.name refer to full extenson_name? Is it same full name defined in policy file?

Ex: Microsoft.Azure.Monitor.AzureMonitorLinuxAgent

Reply:
Correct. extension_to_check is an Extension object, .name refers to the full name. Policy file is expected to have that same full extension name.